### ISSUE TITLE
Question on mathematical equivalence

### ISSUE THREAD

**Issue by nlpfollower** [frankxwang/dpo-prefix-sharing#2]
Hey, really like this idea!

Was experimenting with the mask a bit, and was curious to know whether you tested for mathematical equivalence of the masked and unmasked forward steps for Llama?

I tried running `unmasked = model(prompt + rejected, mask=None)` and then `masked = model(prompt + chosen + rejected, mask=block_mask)`, and then comparing the logits of `unmasked` and `masked[:prompt] + masked[prompt + chosen:]`. However, the logits for the rejected part seem to differ in my experiments, at least when using Llama. The results were the same for a 2D bool mask and sdpa.

It's possible this is an issue with my implementation. I'll see if I can reproduce it in your repo.

**Comment by nlpfollower** [comment]
Okay, it seems like I was able to reproduce this experiment on a fork (https://github.com/nlpfollower/dpo-prefix-sharing/pull/1), running single-rank Llama3.1-8B.

In my original implementation, I tested using the document mask (e.g. https://github.com/pytorch-labs/attention-gym/blob/main/attn_gym/masks/document_mask.py) with packing -- i.e. the boring approach -- and there was still a difference between the logits, though it was significantly smaller compared to the prefix sharing mask.

Do you see any potential issues with my code or this experiment?

**Comment by frankxwang** [comment]
Hi! Sorry for the late response, didn't see this until now. We verified the equivalence by looking at the loss curves of training runs using prefix sharing vs non-prefix sharing, and the trajectories line up almost exactly. As a sanity check, we also evaluated a model trained on UltraFeedback with prefix sharing and sequence packing on MT-Bench, and we found that the performance was very strong (see appendix A of the paper).

When comparing the logits directly, you'll find that the logits between FlexAttention and FlashAttention will be off, but this is to be expected since bfloat16 is not very precise (especially since the magnitude of the logits can be quite large). Ultimately, we find that these precision issues don't really seem to impact training performance.

Let me know if you have any other questions!
